tools/codegen/api/python.py

```diff
@@ -0,0 +1,927 @@
+    'const c10::optional<Tensor>&': 'optionalTensor',
+    'c10::optional<Generator>': 'generator',
+    'Storage': 'storage',
+    'Storage &': 'storage',
+    'const ScalarType &': 'scalartype',
+    'const Device &': 'device',
+    'c10::optional<DimnameList>': 'toDimnameListOptional',
+    'c10::optional<ScalarType>': 'scalartypeOptional',
+    'c10::optional<Layout>': 'layoutOptional',
+    'c10::optional<MemoryFormat>': 'memoryformatOptional',
+    'c10::optional<Scalar>': 'scalarOptional',
+    'c10::optional<IntArrayRef>': 'intlistOptional',
+    'c10::optional<int64_t>': 'toInt64Optional',
+    'c10::optional<bool>': 'toBoolOptional',
+    'c10::optional<double>': 'toDoubleOptional',
+    'c10::optional<ArrayRef<double>>': 'doublelistOptional',
+    'ArrayRef<double>': 'doublelist',
+    'IntArrayRef': 'intlist',
+    'Scalar': 'scalar',
+    'ScalarType': 'scalartype',
+    'Dimname': 'dimname',
+    'DimnameList': 'dimnamelist',
+    'TensorList': 'tensorlist',
+    'int64_t': 'toInt64',
+    'bool': 'toBool',
+    'double': 'toDouble',
+    'std::string': 'string',
+    'c10::optional<std::string>': 'stringOptional',
+}
+
+UNPACK_WITH_SIZE_METHODS = {
+    'TensorList': 'tensorlist_n<{}>',
+    'DimnameList': 'dimnamelist',
+    'IntArrayRef': 'intlist',
+    'c10::optional<IntArrayRef>': 'intlistOptional',
+}
+
+UNPACK_WITH_DEFAULT_METHODS = {
+    'const ScalarType &': 'scalartypeWithDefault',
+    'const Device &': 'deviceWithDefault',
+    'c10::optional<Layout>': 'layoutWithDefault',
+}
+
+# Return RHS expression for python argument using PythonArgParser output.
+# e.g. for arg name 'foo', arg type 'bool', arg_index = 2, returns '_r.toBool(2)'
+def arg_parser_output_expr(
+    arg_index: int, a: PythonArgument, la: Optional[DispatchLambdaArgument]
+) -> PythonArgParserOutputExpr:
+    # The same python signature (and python schema string) is usually
+    # associated with two aten C++ functions: the base version and the
+    # out-place variant. Usually the two functions have the same set of
+    # arguments - of course, except for the output arguments. But in some
+    # cases they might have slightly different C++ argument types -
+    # affected by the 'use_c10_dispatcher' state.
+    #
+    # More specially, 'Tensor?' type can be translated into
+    # either 'const c10::optional<Tensor>&' or 'const Tensor &'.
+    # Unfortunately, this difference can affect how we should access arg
+    # parser output. The former expects '_r.optionalTensor(i)' while the
+    # latter expects '_r.tensor(i)'.
+    #
+    # Because of this subtle difference, we cannot solely use the shared
+    # python signature to determine the RHS expr for both C++ variants.
+    # We could create and use each C++ variant's own python signature,
+    # but we have to fix the argument index difference between the two
+    # python signatures like the old codegen does - and it feels wrong as
+    # technically there is only one shared python signature!
```

Comment:
Yeah - once we've migrated out-place variants to c10_full, the unboxed dispatcher calls (inside the dispatch lambda) will have out args at the end, so the index-shifting can be dispensed with, which should make reuse slightly less awkward.